uefi: Add safe protocol wrapper for EFI_EXT_SCSI_PASS_THRU_PROTOCOL

seijikun opened the pull request:
**Update** [Latest API discussions](https://github.com/rust-osdev/uefi-rs/pull/1589#issuecomment-2791197828)

---

Added a safe wrapper for the extended SCSI Pass Thru protocol.
I did my best to design the API in a way that avoids accidentally using it incorrectly, yet allowing it to operate efficiently.
The `ScsiRequestBuilder` API is designed in a way that should easily make it possible to use it for both `EFI_EXT_SCSI_PASS_THRU_PROTOCOL` and a possible future safe protocol wrapper of `EFI_SCSI_IO_PROTOCOL`.

### Exemplary usage to probe all devices potentially connected to every SCSI channel in the system:

#### Easy variant (io/cmd buffer allocations per request):
```rust
// query handles with support for the protocol (one handle per SCSI controller)
let scsi_ctrl_handles = uefi::boot::find_handles::<ExtScsiPassThru>().unwrap();
// Iterate over scsi controllers with passthru support:
for handle in scsi_ctrl_handles {
    // open protocol for controller
    let scsi_pt = uefi::boot::open_protocol_exclusive::<ExtScsiPassThru>(handle).unwrap();
    // iterate (potential!) devices on the controller
    for device in scsi_pt.iter_devices() {
        // this is not an actual (guaranteed-to-exist) device, but a !!potential!! device.
        // We have to probe it to find out if there is something connected to this chan/target/lun

        // construct SCSI INQUIRY (0x12) request
        let request = ScsiRequestBuilder::read(scsi_pt.io_align())
            .with_timeout(Duration::from_millis(500))
            .with_command_data(&[0x12, 0x00, 0x00, 0x00, 0xFF, 0x00]).unwrap()
            .with_read_buffer(255).unwrap()
            .build();
        // send device through controller to potential device
        if let Ok(response) = device.execute_command(request) {
            println!(
                "SUCCESS HostAdapterStatus: {:?}, TargetStatus: {:?}\r",
                response.host_adapter_status(),
                response.target_status()
            );
            let inquiry_response = response.read_buffer().unwrap();
            println!("ResponseBfr: {:?}\r", inquiry_response);
        } else {
            println!("ERROR - probably not a device\r");
        }
    }
}
```

#### Buffer-reuse API variant:
```rust
[...]
    // open protocol for controller
    let scsi_pt = uefi::boot::open_protocol_exclusive::<ExtScsiPassThru>(handle).unwrap();
    // allocate buffers and reuse amongst drives on this SCSI controller
    // It's important this is not shared across SCSI controllers !! Alignment differs
    let mut cmd_bfr = scsi_pt.alloc_io_buffer(6).unwrap();
    cmd_bfr.copy_from(&[0x12, 0x00, 0x00, 0x00, 0xFF, 0x00]);
    let mut read_bfr = scsi_pt.alloc_io_buffer(255).unwrap();
    // iterate (potential!) devices on the controller
    for device in scsi_pt.iter_devices() {
        // this is not an actual devices, but a !!potential!! device.
        // We have to probe it to find out if there is something connected to this chan/target/lun

        // construct SCSI INQUIRY (0x12) request
        let request = ScsiRequestBuilder::read(scsi_pt.io_align())
            .with_timeout(Duration::from_millis(500))
            .use_command_buffer(&mut cmd_bfr).unwrap()
            .use_read_buffer(&mut read_bfr).unwrap()
            .build();
[...]
```

## Checklist
- [x] Sensible git history (for example, squash "typo" or "fix" commits). See the [Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) guide for help.
- [x] Update the changelog (if necessary)

seijikun commented:
![image](https://github.com/user-attachments/assets/871d4a37-6615-41d8-a2ce-84e5dff285dd)

phip1611 commented on uefi-raw/src/protocol/scsi.rs:
```diff
@@ -80,17 +80,75 @@ newtype_enum! {
+    ///
+    /// A `timeout` value of 0 indicates that the function will wait indefinitely for
+    /// the execution to complete. If the execution time exceeds the specified `timeout`
+    /// (greater than 0), the function will return `EFI_TIMEOUT`.
```
nit: could be `[``TIMEOUT``]` linking to https://docs.rs/uefi-raw/latest/uefi_raw/struct.Status.html#associatedconstant.TIMEOUT

phip1611 commented on uefi/src/helpers/aligned_buffer.rs:
```diff
@@ -0,0 +1,49 @@
+use alloc::alloc::{alloc, dealloc, Layout, LayoutError};
+
+/// Helper class to maintain the lifetime of a memory region allocated with a non-standard alignment.
+/// Facilitates RAII to properly deallocate when lifetime of the object ends.
```
nit: 

"This uses the global Rust allocator under the hood, and thus not necessarily the."

seijikun replied:
I assume you want me to add this line to the comments (?), but I don't know how the sentence should end :sweat_smile:.

phip1611 commented on uefi/src/proto/scsi/mod.rs:
```diff
@@ -0,0 +1,376 @@
+///
+/// This ensures that all buffers used in SCSI commands conform to the alignment requirements of the protocol.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AlignmentError;
```
this struct and change belongs to the previous commit introducing the AlignedBuffer struct I think

phip1611 commented on uefi/src/proto/scsi/mod.rs:
```diff
@@ -0,0 +1,376 @@
+/// Provides a safe and ergonomic interface for configuring SCSI request packets, including timeout,
+/// data buffers, and command descriptor blocks.
+#[derive(Debug)]
+#[repr(transparent)]
```
not sure if we need `repr(transparent)` here; I don't think this type needs to be ABI compatible

phip1611 reviewed (changes requested):
Thanks for your contribution! I left a few remarks. Further, can you please add an integration test for the new functionality?

nicholasbishop commented on uefi-raw/src/protocol/scsi.rs:
```diff
@@ -206,9 +206,9 @@ pub struct ExtScsiPassThruProtocol {
         lun: *mut u64,
     ) -> Status,
-    pub reset_channel: unsafe extern "efiapi" fn(this: *mut Self) -> Status,
+    pub reset_channel: unsafe extern "efiapi" fn(this: *const Self) -> Status,
```
I think stick with `*mut`. It's not a [safety concern](https://github.com/rust-osdev/uefi-rs/blob/main/uefi-raw/api_guidelines.md#mutability) here, but semantically it sounds like an operation where the caller should have exclusive ownership in order to perform a reset. The `uefi` wrapper should then be `&mut`.

seijikun commented:
I hope I addressed all opened review comments now.
CI on x86_64 is green now, including the `AlignedBuffer` unit-test and the integration-test.
For the integration test, I changed the second (FAT32 formatted) disk to SCSI.

Since that change, the qemu of aarch64 integration test runner is hard-crashing in the `Disk I/O 2` test..
Do you have an idea of what this might be?

phip1611 commented on uefi/src/proto/scsi/pass_thru.rs:
```diff
@@ -0,0 +1,239 @@
+    type Item = ScsiDevice<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // no idea why get_next_target_lun() takes a doubleptr ... makes no sense
```
See spec :)
![image](https://github.com/user-attachments/assets/0f5199df-4118-41b7-90ac-a75d9f7e4a06)

seijikun replied:
Yeah, I read that when implementing it. But my testing showed, that the call never returned a new pointer, it instead always wrote the response into the target buffer I gave it. And for that -  a double pointer doesn't make sense.

But now that I think about it ... the spec formulation probably means that the UEFI implementation is **allowed** to return me a new buffer instead of overwriting my buffer. I'll fix my implementation to honor that.

phip1611 reviewed (approved):
This is a good PR with a comprehensive improvement! Thanks! LGTM

As this PR is a little bigger than others, I'd like  @nicholasbishop to leave a second approval

phip1611 commented on uefi/src/helpers/aligned_buffer.rs:
```diff
@@ -0,0 +1,79 @@
+
+    #[test]
+    fn test_allocation_alignment() {
+        for request_alignment in [1, 2, 4, 8, 16, 32, 64, 128] {
```
nit: please also test the edge cases alignment=0 and size=zero.

nicholasbishop commented on uefi-raw/src/protocol/scsi.rs:
```diff
@@ -150,7 +208,7 @@ pub struct ExtScsiPassThruProtocol {
     ) -> Status,
     pub reset_channel: unsafe extern "efiapi" fn(this: *mut Self) -> Status,
     pub reset_target_lun:
-        unsafe extern "efiapi" fn(this: *mut Self, target: *const u8, lun: u64) -> Status,
```
This one also feels "logicially mut" to me, can we make this mut and ditto for the corresponding `uefi` wrapper?

seijikun replied:
I can make the raw API mut, **but**...:

The entity in the safe `uefi` wrapper that corresponds to this funcitonality is `ScsiDevice<'a>` in function `ScsiDevice::reset()`.
And the only way to get an instance of `ScsiDevice<'a>` is through the device iterator: `ExtScsiPassThru::iter_devices()`.

How would you suggest to make this mut? Have a second `ExtScsiPassThru::iter_devices_mut()`?

nicholasbishop replied:
I think `ScsiDevice::execute_command` should also be `&mut`. So maybe there should only be `iter_devices_mut`, and drop `iter_devices` (since `iter_devices` indicates you will need to probe with `execute_command` anyway). Would that work, or do you see problems in API usage if we require exclusive access?

nicholasbishop replied:
Oh, and a general note: even if we decide to make the `uefi` interface `&` instead of `&mut`, we can still use `*mut` in the raw interface. Making a mut pointer from a non-mut reference is allowed, and doesn't affect soundness in and of itself.

seijikun replied:
I just rewrote the API to use mutable and I ran into some ugly problems:

When you call `iter_devices_mut()` to produce `ScsiDevice`s, you mutably borrow the protocol (`ExtScsiPassThru`) itself for the lifetime of the generated `ScsiDevice`. Thus, while a device exists, you can't interact with the protocol itself. That produces all kinds of usability issues.

```rust
for mut device in scsi_pt.iter_devices_mut() {
            let request = ScsiRequestBuilder::read(scsi_pt.io_align())
                .with_timeout(Duration::from_millis(500))
                ...
```
![image](https://github.com/user-attachments/assets/81a91dc4-b817-4ae0-b8d8-9610ae5dce65)

phip1611 replied:
Thanks for experimenting and analyzing this. I don't have the capacity right not unfortunately to provide good guidance. @nicholasbishop  can you help, please?

nicholasbishop commented on uefi/src/helpers/aligned_buffer.rs:
```diff
@@ -0,0 +1,97 @@
+pub struct AlignedBuffer {
+    ptr: *mut u8,
+    layout: Layout,
+}
```
nit: add a blank line after a `}` that closes a struct or function (so, also add blank line separating each method in the `impl` below)

nicholasbishop commented:
I'm looking through this in more detail now :) Mind moving the first commit ("uefi-raw: Add documentation to ScsiIoScsiRequestPacket") to a new PR? Can merge that separately from the rest of the changes.

seijikun commented:
Done: https://github.com/rust-osdev/uefi-rs/pull/1593

nicholasbishop commented on uefi/src/helpers/mod.rs:
```diff
@@ -24,6 +24,12 @@ use crate::Result;
 
+#[cfg(feature = "alloc")]
+mod aligned_buffer;
+#[doc(hidden)]
```
Why hidden?

nicholasbishop commented on uefi/src/helpers/aligned_buffer.rs:
```diff
@@ -0,0 +1,97 @@
+/// Note: This uses the global Rust allocator under the hood.
+#[allow(clippy::len_without_is_empty)]
+#[derive(Debug)]
+pub struct AlignedBuffer {
```
For SCSI, does the alignment ever exceed 4096? I'm wondering if it would make more sense for callers to allocate memory with `boot::allocate_pages`, which are always 4K aligned. We might not need `AlignedBuffer` in that case. (It might still be a helpful interface either way, I just want to see if a simpler solution can work.)

seijikun replied:
Main idea behind the AlignedBuffer struct was, that I don't have to think if `dealloc`-ing a buffer I allocated myself.
I disliked having the user copy over the `io_align` myself - but I'd be too scared to make assumptions as to what can be expected as possible values.

seijikun replied:
Instead of letting a user construct the request builder freely, we could instead add something like:
```rust
pub enum ScsiRequestDirection { READ, WRITE, BIDIRECTIONAL }
```
And then add the following method to `ExtScsiPassThru` for starting new requests:
```rust
pub fn start_request(&self, direction: ScsiRequestDirection) -> ScsiRequestBuilder {}
```
That would then avoid the user ever coming into direct contact with `io_align`.


So something like this:
```rust
let scsi_pt = uefi::boot::open_protocol_exclusive::<ExtScsiPassThru>(handle).unwrap();
let request = ScsiRequestBuilder::read(scsi_pt.io_align())
    .with_timeout(Duration::from_millis(500))
    .build();
```
becomes:
```rust
let scsi_pt = uefi::boot::open_protocol_exclusive::<ExtScsiPassThru>(handle).unwrap();
let request = scsi_pt.start_request(ScsiRequestDirection::READ)
    .with_timeout(Duration::from_millis(500))
    .build();
```

seijikun commented:
@nicholasbishop I think I might have a nice solution to the mutable / immutable problem. Please have a look at https://github.com/rust-osdev/uefi-rs/pull/1589/commits/5a44ef4839b465d0e779f49929ba8a94bf3b7270

Usage looks like this:
```rust
let scsi_ctrl_handles = uefi::boot::find_handles::<ExtScsiPassThru>().unwrap();
for handle in scsi_ctrl_handles {
    //     v  does not require mutable
    let scsi_pt = uefi::boot::open_protocol_exclusive::<ExtScsiPassThru>(handle).unwrap();
    //  ScsiDevices returned from the device iterator are owned structs (no &ref, no &mut ref)
    //   v requires mutable            v does not require mutable
    for mut device in scsi_pt.iter_devices() {
        println!("- Disk: {:?} | {}\r", device.target(), device.lun());

        let request = ScsiRequestBuilder::read(scsi_pt.io_align())
            .with_timeout(Duration::from_millis(500))
            .use_command_data(&[0x06, 0x00, 0x00, 0x00, 0x00]).unwrap()
            .with_read_buffer(255).unwrap()
            .build();
        //             v requires device to be mutable
        let result = device.execute_command(request).unwrap();
    }
}
```

What do you think about this?

nicholasbishop commented:
I'm wondering if some of the lifetime complexity could be reduced by moving the methods of `ScsiDevice` to `ExtScsiPassThru`. So instead of `ScsiDevice` being a "smart" object that you can call `execute_command` (and `reset`) on, you would call `execute_command` on the protocol, and pass the `ScsiTargetLun` as an argument.

Would that work? If so, are there downsides I'm not thinking of?

seijikun commented:
Yes, that would eliminate the lifetime problems. If you prefer that, I can refactor it.

nicholasbishop commented:
Yes, as long as it doesn't cause other issues I think that would be a good refactor. Fewer lifetimes, and avoiding internal pointers and Phantom markers feels worth it to me, to make it less likely that the code is accidentally unsound.

seijikun commented:
*grr*
I just did the refactoring, and that architecture produces the same lifetime problems.
I can not send commands within an iterator, because the device iterator borrows the protocol.

nicholasbishop commented:
Ah right. What about collecting the iterator into a `Vec` first?

seijikun commented:
Yes, that works. Although it's an additional allocation. On the other hand, we require `alloc` for this anyway.
Your call.

seijikun commented:
My personal favorite stays the smart object design I suggested. It has several advantages:
- By far the best usability
- It avoids an unnecessary allocation (no `iter_devices().collect()`)
- It actually makes sense, in that:
    - it does not allow you to borrow the protocol instance mutably while you have `ScsiDevice` instances alive. So you can e.g. not reset the entire channel with `ScsiDevice`'s borrowing the protocol immutably
    - it's not possible to accidentally use a `ScsiDevice` instance with another `ExtScsiPassThru` protocol instance

I just pushed a version without internal pointer and without `PhantomData` usage - thus alleviating some of your concerns with it.

nicholasbishop commented:
FYI, I'll return to looking at this soon -- I've been a way for a few days so lots to catch up on at work :)

nicholasbishop commented:
Mind pulling AlignedBuf into a separate PR? I think I'm sold on the usefulness of that addition to the API, and the need for aligned buffers comes up in lots of places, so I'd like to review that independently.

phip1611 commented on uefi/src/proto/scsi/pass_thru.rs:
```diff
@@ -0,0 +1,269 @@
+/// This protocol allows communication with SCSI devices connected to the system,
+/// providing methods to send commands, reset devices, and enumerate SCSI targets.
+///
+/// This API offers a safe and convenient, yet still low-level interface to ATA/NVMe/SCSI devices.
```
```suggestion
/// This API offers a safe and convenient, yet still low-level interface to SCSI devices.
```

seijikun commented:
I think the API discussion has drifted off a bit and is somewhat widespread throughout various "threads" in this issue.
As an attempt to get this PR back on track, I'll try to quicly summarize the options/variants I see for the device interaction API and lay out their potential problems / advantages:

---

### 1) Intelligent Device Objects: Everything const
- Device Iterator yields intelligent objects that directly provide APIs for interaction with devices on the bus
- All interaction requires only const references (on protocol and on devices) - including actions like reset_channel() to reset the entire bus

**API**
```rust
ExtScsiPassThru::reset_channel(&self);
// returned ScsiDevice is owned (not a reference), but internally borrows the underlying raw protocol
ExtScsiPassThru::iter_devices(&self) -> Iterator<Item = ScsiDevice>;

ScsiDevice::reset(&self);
ScsiDevice::execute_command(&self, cmd: ScsiCommand);
```

Usability wise (w.r.t. to flexibility for weird use-cases), probably the best option.
This would allow mixing bus resets with device interactions without having to call the iterator again.
Allowing this is not unsafe, it's just that all subsequent device interaction (if the device vanished from the bus due to the reset) returns errors.
I don't know if there is a use-case where a mixed-in bus reset makes sense.

**Usage**
```rust
let proto: ExtScsiPassThru;
for device in proto.iter_devices() {
    device.execute_command(..);
    proto.reset_channel(); // << allows bus-reset mixed with device interaction
}
// also allows:
let devices = proto.iter_devices().collect();
devices[0].reset();
devices[1].reset();
```

---

### 2) Intelligent Device Objects: Actions mut
- Protocol-wide actions (bus reset) require a mutable protocol reference.
- The device iterator only captures the protocol const
- Device-specific actions (device reset, execute command) require a mutable device reference.

**API**
```rust
ExtScsiPassThru::reset_channel(&mut self);
// returned ScsiDevice is owned (not a reference), but internally borrows the underlying raw protocol
ExtScsiPassThru::iter_devices(&self) -> Iterator<Item = ScsiDevice>;

ScsiDevice::reset(&mut self);
ScsiDevice::execute_command(&mut self, cmd: ScsiCommand);
```

Usability wise, very close to all-const option.
Only difference is that it forbis bus resets while a `ScsiDevice` object lives.
Forbidding this makes sense if you assume, that a device doesn't reappear after a bus reset - which is IMHO a valid possibility.

**Usage**
```rust
let proto: ExtScsiPassThru;
for mut device in proto.iter_devices() {
    device.execute_command(..);
}
// also allows:
let devices = proto.iter_devices().collect();
devices[0].reset();
devices[1].reset();
```

---

### 3) Device Handle: Actions at protocol mut
- All interaction (either bus-wide, or device specific) happen through methods in `ExtScsiPassThru` that require a mutable reference.
- Device iterator yields dumb handles that have no functionality themselves. They get passed to the device-specific methods in `ExtScsiPassThru` for device interaction.

**API**
```rust
ExtScsiPassThru::reset_channel(&mut self);
// Returned ScsiDevice objects are only handles passed back to ExtScsiPassThru methods
// The objects don't keep a reference to the protocol
ExtScsiPassThru::iter_devices(&self) -> Iterator<Item = ScsiDevice>;

ExtScsiPassThru::reset_device(&mut self, dev: ScsiDevice);
ExtScsiPassThru::execute_command(&mut self, dev: ScsiDevice, cmd: ScsiCommand);
```

Usability-wise, this variant is worse than the other two, because:
  - The API discoverability is lower, because the ScsiDevice objects are only handles
  - The iterator borrows the Protocol, so you can't easily interact with devices while iterating, you always need to allocate
However: It allows resetting the bus while ScsiDevice instances live, just like the all-const variant.
And: It allows taking the `ScsiDevice` handle from one `ExtScsiPassThru` instance and pass it to another, which is at least as "wrong" as what can happen with a bus reset.

**Usage**
```rust
let proto: ExtScsiPassThru;
for device in proto.iter_devices() {
    // not possible! Iterator borrows protocol immutably, interaction requires mut
    device.execute_command(device, ..);
}
// you have to allocate:
let devices = proto.iter_devices().collect();
for device in devices {
    proto.execute_command(device, ..);
    proto.reset_channel(); // << allows bus-reset mixed with device interaction
}
// also allows:
device.execute_command(device[0], ..);
device.execute_command(device[1], ..);
```

---

IMHO, **2) Intelligent Device Objects: Actions mut** is the best option. That's the one that is implemented at the moment.
Does not make weird use-cases like the mixed-in bus-reset outright impossible, just makes it less convenient.

I don't want to create any pressure, but I'd be happy if we could decide on one variant of these soon (or a concrete counterproposal I can implement).

phip1611 commented:
@seijikun Thanks for the write-up. That is definitely very helpful.

I'm also in favor of `2) Intelligent Device Objects: Actions mut`.

In case the test of time shows that this was a bad design, we can just make a breaking chance (we are still not v1.0.0).

One more question: Do you see a possibility to abstract over all passthru storage device protocols to not reimplement the same API multiple times?

seijikun commented:
> One more question: Do you see a possibility to abstract over all passthru storage device protocols to not reimplement the same API multiple times?

I think the largest duplication effort is in the RequestBuilders, and that one can't be abstracted away because of how different these protocols are.
Especially NVMe is also very different in semantics. While one `NvmePassThru` instance represents one NVMe drive, one instance of `ExtScsiPassThru` and `AtaPassThru` represents a controller with potentially connected drives, or splitters and even more drives. (NVMe namespaces are like partitions of the same physical drive)

So if at all, I think we could only put SCSI and ATA in the same abstraction - but even they differ quite a lot in their implementation. Compare e.g. the iterator implementation, which is much more contrived with ATA, because of the splitter semantics.
I doubt we could save in the department of raw code lines, sadly.

phip1611 commented:
> So if at all, I think we could only put SCSI and ATA in the same abstraction - but even they differ quite a lot in their implementation.

No problem, that's fine. We do not need abstractions at any cost

nicholasbishop commented:
Thanks for the detailed writeup! And sorry for the delay in getting you my review, been quite busy recently :)

I agree with your analysis, let's stick with option 2 (and as Philipp said, if we need to change it in the future we can).

nicholasbishop commented:
Merged, thanks for your patience!

A couple suggestions to reduce delays for the future:
1. As you noted, discussion of the design ended up playing out across a number of PRs. Since the end design for nvme/scsi/ata passthrough would probably end up similarly, it'd probably have been a little smoother to just put up one of these PRs at a time; get it reviewed and merged before putting up the next. (This is just a suggestion for future work, no need to close any of your current PRs).
2. I'm a big fan of small PRs that build towards an end goal. It's much easier to review a 100 line change than a 1000 line change, and the review can be more thorough with a smaller PR. I recognize in this case you're adding a whole new API and tests, so it can be hard to find a good place to split. In that case, splitting into small commits within the PR can be helpful. https://google.github.io/eng-practices/ has some nice guidance for code authors and reviewers that I like.

Hope these are helpful, and thanks again for all your work on these protocols.
